htlcswitch/link.go

```diff
@@ -2539,28 +2542,43 @@ func (l *channelLink) UpdateForwardingPolicy(newPolicy ForwardingPolicy) {
 func (l *channelLink) CheckHtlcForward(payHash [32]byte,
 	incomingHtlcAmt, amtToForward lnwire.MilliSatoshi,
 	incomingTimeout, outgoingTimeout uint32,
+	inboundFee InboundFee,
 	heightNow uint32, originalScid lnwire.ShortChannelID) *LinkError {
 
 	l.RLock()
 	policy := l.cfg.FwrdingPolicy
 	l.RUnlock()
 
-	// Using the amount of the incoming HTLC, we'll calculate the expected
-	// fee this incoming HTLC must carry in order to satisfy the
-	// constraints of the outgoing link.
-	expectedFee := ExpectedFee(policy, amtToForward)
+	// Using the outgoing HTLC amount, we'll calculate the outgoing
+	// fee this incoming HTLC must carry in order to satisfy the constraints
+	// of the outgoing link.
+	outFee := ExpectedFee(policy, amtToForward)
+
+	// Then calculate the inbound fee that we charge based on the sum of
+	// outgoing HTLC amount and outgoing fee.
+	inFee := inboundFee.CalcFee(amtToForward + outFee)
+
+	// Add up both fee components. It is important to calculate both fees
+	// separately. An alternative way of calculating is to first determine
+	// an aggregate fee and apply that to the outgoing HTLC amount. However,
+	// rounding may cause the result to be slightly higher than in the case
+	// of separately rounded fee components. This potentially causes failed
+	// forwards for senders and is something to be avoided.
+	expectedFee := inFee + int64(outFee)
 
 	// If the actual fee is less than our expected fee, then we'll reject
 	// this HTLC as it didn't provide a sufficient amount of fees, or the
 	// values have been tampered with, or the send used incorrect/dated
 	// information to construct the forwarding information for this hop. In
-	// any case, we'll cancel this HTLC. We're checking for this case first
-	// to leak as little information as possible.
-	actualFee := incomingHtlcAmt - amtToForward
-	if incomingHtlcAmt < amtToForward || actualFee < expectedFee {
+	// any case, we'll cancel this HTLC.
+	actualFee := int64(incomingHtlcAmt) - int64(amtToForward)
+	if actualFee < expectedFee {
```

Comment:
I think you could change this to `if incomingHtlcAmt < amtToForward || actualFee < expectedFee` as a belt-and-suspenders check to ensure that the node is not losing money due to an overall negative fee.

Reply:
Indeed checking it when setting the policy may get a bit complicated. By checking here the node as you say doesn't lose money but will just be penalized by the sender.

But is it really not desired to allow nodes to set a negative (total) fee as a way to incentivize traffic in a certain direction? Currently they may execute rebalance payments which also makes them lose money.

Reply:
>I'm not familiar with the mission control code, but being penalized by the sender definitely doesn't sound good. 

If the sender receives a fee insufficient failure repeatedly, despite the (negative) fee calculation being correct, the sender will penalize that routing node. I do think that it is reasonable to say that in most cases this is preferable over  seeing channel balance drained because of an accidental negative fee setting.

>If we go with the disallow negative total fee approach, what if the path-finding code ensured that the incoming amount was always greater than the outgoing amount?

Could do that, but other implementations may differ.

Overall I think that failing forwards with a negative fee is a good idea in combination with documenting that negative fee policies set by the user won't be honored. Can always relax this later on.

Reply:
Restored `incomingHtlcAmt < amtToForward` condition.

Reply:
A good idea to not allow negative forward fees for the time being, although I think it's a good idea as well. But I think we need more research on the necessary conditions to the fees on each channel pair, so that a noderunner doesn't lose money in the end.